File: 1001-1100/1074-Number of Submatrices That Sum to Target/1074-Number of Submatrices That Sum to Target.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
        vector<vector<int>> presum(matrix.size(), vector<int>(1 + matrix[0].size()));
        for (int i = 0; i < matrix.size(); ++i) {
            for (int j = 0; j < matrix[0].size(); ++j) {
                presum[i][j + 1] = presum[i][j] + matrix[i][j];
            }
        }

        int total = 0;
        for (int j = 0; j < matrix[0].size(); ++j) {
            for (int i = 0; i <= j; ++i) {
                int curr = 0;
                unordered_map<int, int> table;
                table[0] = 1;
                for (int k = 0; k < matrix.size(); ++k) {
                    curr += presum[k][j + 1] - presum[k][i];
                    total += table[curr - target];
                    ++table[curr]; 
                }
            }
        }

        return total;
    }
};

File: 1001-1100/1074-Number of Submatrices That Sum to Target/1074-Number of Submatrices That Sum to Target.py
import collections
class Solution:
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        presum = [[0] * (len(matrix[0]) + 1) for _ in range(len(matrix))]
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                presum[i][j + 1] = presum[i][j] + matrix[i][j]

        total = 0
        for j in range(len(matrix[0])):
            for i in range(j + 1):
                curr = 0
                table = collections.Counter([0])
                for k in range(len(matrix)):
                    curr += presum[k][j + 1] - presum[k][i]
                    total += table[curr - target]
                    table[curr] += 1

        return total
